### PR TITLE
Cloudflare HTTPS Guide

### PR DESCRIPTION
**Pull request by joshkoenig**
At long last!

### CODE REVIEW

**Comment by davidstrauss** [comment]
It might be worth noting that there is the option of using *strict* SSL to Pantheon with a CloudFlare enterprise account configured to expect Pantheon's *.pantheon.io certificate. Or maybe that can wait for version two of this guide.

**Comment by davidstrauss** [comment]
I would also tone this down:

> However, this free option still provides solid security for your website: the traffic is fully encrypted end-to-end.

It's nice that it's encrypted. That prevents more casual eavesdropping, but it won't stop a serious MitM attack.

**Comment by atjuch** [comment]
Josh, do you want this doc to pretty much replace this?

https://pantheon.io/docs/articles/wordpress/add-cloudflare-free-ssl-to-wordpress-sites/

Since you are having the rewrite rules added instead of depending on a
HTTPS plugin.

**Comment by joshkoenig** [comment]
@atjuch I think we should probably reference the doc which that was adapted from.

@davidstrauss Not sure about this:

bq. It might be worth noting that there is the option of using strict SSL to Pantheon with a CloudFlare enterprise account configured to expect Pantheon's *.pantheon.io certificate.

According to Trey what they did for that was a PoC, and not available on-demand, yet. I'd rather wait until it's a confirmed first-class product feature. However, I can add that "strict" mode is achievable if you obtain your own cert, and have the budget to set that up on Pantheon. 

My positive note about end-to-end encryption is to contrast us against the "flexible" mode, which is currently what a lot of people use.

**Comment by rachelwhitton** [comment]
Awesome guide @joshkoenig! Thanks! Can you review/edit the author info I added for you? 

@nataliejeremy testing complete! +1

**Comment by joshkoenig** [comment]
@rachelwhitton +1 LGTM

**Comment by suzannealdrich** [comment]
@joshkoenig Shouldn't we also be recommending enabling Authenticated Origin Pulls, in order to prevent bypassing CloudFlare security measures? See https://support.cloudflare.com/hc/en-us/articles/204899617

**Comment by suzannealdrich** [comment]
@joshkoenig My WP sites threw errors when re-defining the WP_HOME and WP_SITEURL constants, and I ended up modifying the Pantheon supplied wp-config.php configuration for these with this:
```
      define('WP_HOME', 'https://' . $_SERVER['HTTP_HOST']);
      define('WP_SITEURL', 'https://' . $_SERVER['HTTP_HOST']);
```

**Comment by joshkoenig** [comment]
I believe that Authenticated Origin Pulls would require _us_ to validate _their_ cert when the site is being requested. That doesn't exist at the moment, and would take some deeper integration.

**Comment by joshkoenig** [comment]
@suzannealdrich can you show me where changing the `WP_HOME` and `WP_SITE_URL` constants to be specific caused errors?

**Comment by rachelwhitton** [comment]
I didn't receive any errors when changing the constants on my test WP site. All environments were successfully encrypted.

**Comment by suzannealdrich** [comment]
@joshkoenig It was my misunderstanding on Authenticated Origin Pulls. We would need to add this configuration to our application server nginx configuration for that to properly authenticate traffic is coming from CloudFlare: https://support.cloudflare.com/hc/en-us/articles/204494148-Setting-up-NGINX-to-use-TLS-Authenticated-Origin-Pulls

**Comment by joshkoenig** [comment]
Per Suzanne - let's add:

```
// CloudFlare Connecting IP
if (isset($_SERVER['HTTP_CF_CONNECTING_IP'])) {
  $_SERVER['REMOTE_ADDR'] = $_SERVER['HTTP_CF_CONNECTING_IP'];
}
```

**Comment by suzannealdrich** [comment]
https://gist.github.com/suzannealdrich/bb693a63fcab375024b5

File: source/docs/guides/ssl-with-cloudflare.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    if (isset($_ENV['PANTHEON_ENVIRONMENT'])) {
+      if ($_ENV['PANTHEON_ENVIRONMENT'] === 'dev') {
+        $domain = 'sandbox.mysite.com';
+      }
+      if ($_ENV['PANTHEON_ENVIRONMENT'] === 'test') {
+        $domain = 'staging.mysite.com';
+      }
+      if ($_ENV['PANTHEON_ENVIRONMENT'] === 'live') {
+        $domain = 'www.mysite.com';
+      }
```
Comment (joshkoenig) [inline source/docs/guides/ssl-with-cloudflare.md]: This will break multidev environments.

File: source/docs/guides/ssl-with-cloudflare.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+When you first register, CloudFlare will direct you immediately towards migrating your domain name. It's best to go ahead with this and to take the time to use CloudFlare as your DNS provider. Their handy wizard will help you port over your existing settings.
+
+## DNS Configuration
+
+In addition to gratis HTTPS service, content distribution, and some nifty security features, CloudFlare also offers the ability to use a CNAME for the "root" domain for your site through something they call [CNAME Flattening](https://blog.cloudflare.com/introducing-cname-flattening-rfc-compliant-cnames-at-a-domains-root/).
+
+We recommend you take advantage of this feature as it frees you up from being tied to a single IP address, which is inherently risky. Here are some example DNS settings:
+
+![Example DNS Settings](/source/docs/assets/images/cloudflare-dns.png)
```
Comment (joshkoenig) [inline source/docs/guides/ssl-with-cloudflare.md]: Note that not everyone may have special domain names for dev and test.

**Comment by rachelwhitton** [comment]
@joshkoenig can you review [6e85f9d](https://github.com/pantheon-systems/documentation/commit/6e85f9d5539a442c070ce3b04077bcb01555554d)?

File: source/docs/guides/ssl-with-cloudflare.md (hunk 1 above)
Comment (joshkoenig) [inline source/docs/guides/ssl-with-cloudflare.md]: Resolved.

**Comment by joshkoenig** [comment]
Hey @rachelwhitton actually turns out that we didn't need that `REMOTE_ADDR` bit. Verified in my set-up, and re-worded.

This is now complete!

**Comment by rachelwhitton** [comment]
Sweet! I'll merge and deploy now :) Thanks @joshkoenig!

**Pull request merged**